_published/0013-authentication.md

```diff
@@ -0,0 +1,165 @@
+
+#### The ID Token
+A OIDC type ID-token can be used for authentication. The custom claims
+(requires an IDP with support for RBAC) specified within the ID-token can
+then be used to specify access levels. To keep the scope manageable the task
+of dividing up and refining access levels and permission will be left to a
+later date. For this RFC all kinds of read permissions won't require valid
+ID-tokens, but all write type requests will require valid ID-tokens that
+have been signed by a pre-approved issuer.
+The OIDC protocol defines a token called an
+[ID Token](https://auth0.com/docs/tokens/id-tokens/id-token-structure).
+The ID token is signed by the source issuer. This signature is then used to
+validity and authenticity of the token. See
+[validating an OIDC ID-Token](https://connect2id.com/blog/how-to-validate-an-openid-connect-id-token).
+
+#### Sessions
+The wharf-api is currently stateless and does not use sessions. It is
+possible to either include the ID token in every request to the server or to
+start using sessions to store the ID token between requests.
+
+#### Local development vs production environments
+For the initial release of the feature the configuration will be tailored to
+using azure as an IDP. These settings will as necessary be exposed as
+environment variables to each respective docker image.
+
+
+#### Choice of Client
+
+The OIDC standard defines a
+[list of certified clients](https://openid.net/developers/certified/)
+that can help simplify the implementation of ID Token retrieval.
+
+For the backend a full client will not be required. As the tokens just need
+to be validated and not requested. Validation can be done offline by caching
+the issuers' signature for future validations (OBS! The signature key may
+rotate on a timely basis).
+
+## Compatibility
+
+Expect no compatability between versions of wharf without authentication to
+versions of wharf with authentication. There is a potential threat in leaving
+the system backwards compatible in that a non-secured version of the api could
+run together in an otherwise secure cluster.
+
+We should aim to pick a solution where any IDP supporting OIDC and RBAC can be
+used. Locking ourselves to a single IDP when a standard such as OIDC exists
+seems like a  waste opportunity. Keeping the choice of IDP open will leave
+more room for choice and flexibility in the future.
+
+## Alternative solutions
+
+There are many types of identity verification systems.
+This RFC assumes that OIDC will be used. Within OIDC there are multiple
+different [flow definitions](https://developer.okta.com/docs/concepts/oauth-openid/#choosing-an-oauth-2-0-flow).
+We should primarily use `codeflow with pkce` as this is presumed to be the
+most secure, and best compatible at the time of writing.
+
+## Future possibilities
+
+We may want to implement multiple different IDP sources. This could allow
+login from different sources than the Iver-Azure-AD. A sample situation
```

Comment:
No internal references allowed, sorry

```suggestion
login from different sources than a single Azure-AD. A sample situation
```

This is not written down somewhere, except for this comment here and now, but we got directive to not make any company nor product references throughout the project, not even to Iver itself. Partially to make it seem more open and inviting, but more importantly as a hard line so we don't accidentally slip out references to confidential products or partners.

Iver relying on an Azure AD is not a peice of information that will necessarily collapse the company if it gets into the wrong hands, but it has been declared that we still should refrain from referencing the company where it's not necessary.

I can agree that it's kind of a flimsy rule, but I like it for the fact that it makes us less potent in accidentally exposing confidential or dangerous information.

Reply:
I understand the principals and values in this case the iver ad is public info if not explicitly written then atleast through dns definitions. In this case its the only motivator i know for picking azure ad the IDP. In all other aspects i believe auth0/okta is a better choice in wharf's case.